### PR TITLE
LOG-1267: Loki output implementation.

### PR DESCRIPTION
**Pull request by alanconway**
Issue: https://issues.redhat.com/browse/LOG-1267
Enhancement: https://github.com/openshift/enhancements/blob/master/enhancements/cluster-logging/forward-to-loki.md

Added Loki receiver for functional and e2e tests, using Loki single-process mode.

NOTES:

Added "log_type" field to the log record: "application", "infrastructure", "audit".
The log type is related but not the same as the elasticsearch index,
which has ES-specific prefix/suffix and can be changed by the ES output.

Uses standard go testing.T tests. More compact than Ginkgo tests.
Demonstrates similar before/after, variable scope, and naming as Ginkgo tests.

/assign @jcantrill
/cc @periklis

### CODE REVIEW

File: pkg/apis/logging/v1/output_types.go
Comment (periklis) [inline pkg/apis/logging/v1/output_types.go:158]: nit. Awesome introduction on validation of label keys. If you want to give API docs readers a full insight, I recommend to leave 1 or 2 sentences that the user selection is mandated on loki side by the rate limit config. Precisely these two flags mandate if logs are accepted due to label keys/values length:
```
# Maximum length of a label name.
# CLI flag: -validation.max-length-label-name
[max_label_name_length: <int> | default = 1024]

# Maximum length of a label value.
# CLI flag: -validation.max-length-label-value
[max_label_value_length: <int> | default = 2048]
```
Source: https://grafana.com/docs/loki/latest/configuration/#limits_config

File: pkg/apis/logging/v1/output_types.go
Comment (periklis) [inline pkg/apis/logging/v1/output_types.go:143]: General question, are we omitting the ability for user to set the Tenant ID for the simple forwarding case in custom managed Loki installation. If yes, why?

**Comment by jcantrill** [comment]
@alanconway  please link the needed JIRA to this task for tracking

File: pkg/apis/logging/v1/output_types.go
Comment (alanconway) [reply pkg/apis/logging/v1/output_types.go:143]: I think I misunderstood a conversation we had. We talked about it being sufficient to use the URL path for our own managed Loki. I assumed that extended to all Loki installations. Should I re-instate that ability for user-managed Lokis?

Comment (periklis) [reply pkg/apis/logging/v1/output_types.go:143]: Yes user-managed Loki's should support the TenantID field.

Comment (alanconway) [reply pkg/apis/logging/v1/output_types.go:143]: /hold
Restoring tenantID field.

File: pkg/generators/forwarding/fluentd/loki_conf.go
Comment (4n4nd) [inline pkg/generators/forwarding/fluentd/loki_conf.go:11]: I don't think we should keep `kubernetes.pod_name` as a default label. Since openshift/k8s clusters can have constantly changing pod names this might lead to a high cardinality and reduce querying performance ([Docs](https://grafana.com/blog/2020/04/21/how-labels-in-loki-can-make-log-queries-faster-and-easier/#cardinality)).

File: pkg/apis/logging/v1/output_types.go
Comment (periklis) [inline pkg/apis/logging/v1/output_types.go:149]: As mentioned on the enhancement doc, TenantKey should not be a key on one of the labels but rather an arbitrary string that forwarder users can pick based on their needs.

**Comment by alanconway** [comment]
/unhold
tenantID restored, added tag for uniqueness of non-container logs.
ready for internal trials, may be some changes before release.

**Review by periklis — APPROVED** [review]
LGTM

**Review by periklis — CHANGES_REQUESTED** [review]
@alanconway  This needs a `make bundle` before merging.

**Comment by alanconway** [comment]
@periklis need final review and lgtm.
I had to make some changes to the label set to avoid out-of-order problems.
Explained in the  updated proposal: https://github.com/openshift/enhancements/pull/842 (also could use an LGTM)
Thanks!

**Comment by alanconway** [comment]
/hold
update log_type to log_source based on  https://issues.redhat.com/browse/LOG-1592

**Pull request merged**
